common/flatpak-instance.c

```diff
@@ -428,6 +442,131 @@ flatpak_instance_get_instances_directory (void)
 }
 
+/*
+ * Return the directory in which we create a subdirectory per
+ * concurrently running Flatpak app-ID to store app-specific data that
+ * is common to all instances of the same app.
+ *
+ * This directory is not shared with Flatpak apps, and we rely on this
+ * for the sandbox boundary.
+ *
+ * This is currently the same as the
+ * flatpak_instance_get_instances_directory(). We can distinguish between
+ * instance IDs and app-IDs because instances are integers, and app-IDs
+ * always contain at least one dot.
+ */
+char *
+flatpak_instance_get_apps_directory (void)
+{
+  return flatpak_instance_get_instances_directory ();
+}
+
+/*
+ * @app_id: $FLATPAK_ID
+ * @lock_fd_out: (out) (not optional): Used to return a lock on the
+ *  per-app directories
+ * @lock_path_out: (out) (not optional): Used to return the path to the
+ *  lock file, suitable for bind-mounting into the container
+ *
+ * Create a per-app directory and take out a lock on it.
+ */
+gboolean
+flatpak_instance_ensure_per_app_dir (const char *app_id,
+                                     int *lock_fd_out,
+                                     char **lock_path_out,
+                                     GError **error)
+{
+  glnx_autofd int lock_fd = -1;
+  g_autofree char *lock_path = NULL;
+  g_autofree char *per_app_parent = NULL;
+  g_autofree char *per_app_dir = NULL;
+  struct flock non_exclusive_lock =
+  {
+    .l_type = F_RDLCK,
+    .l_whence = SEEK_SET,
+    .l_start = 0,
+    .l_len = 0
+  };
+
+  g_return_val_if_fail (app_id != NULL, FALSE);
+  g_return_val_if_fail (lock_fd_out != NULL, FALSE);
+  g_return_val_if_fail (*lock_fd_out == -1, FALSE);
+  g_return_val_if_fail (lock_path_out != NULL, FALSE);
+  g_return_val_if_fail (*lock_path_out == NULL, FALSE);
+  g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
+
+  per_app_parent = flatpak_instance_get_apps_directory ();
+  per_app_dir = g_build_filename (per_app_parent, app_id, NULL);
+  lock_path = g_build_filename (per_app_dir, ".ref", NULL);
+
+  if (g_mkdir_with_parents (per_app_dir, 0700) != 0)
+    return glnx_throw_errno_prefix (error,
+                                    _("Unable to create directory %s"),
+                                    per_app_dir);
+
+  /* Take a file lock inside the shared directory, and hold it during
+   * setup and in bwrap. We never delete the directory itself, or the
+   * lock file that it contains (that would defeat the locking scheme).
+   * Anyone cleaning up other members of per_app_dir must first verify
+   * that it contains the lock file .ref, and take out an exclusive lock
+   * on it. */
+  lock_fd = open (lock_path, O_RDWR | O_CREAT | O_CLOEXEC, 0600);
+
+  /* As with the per-instance directories, there's a race here, because
+   * we can't atomically open and lock the lockfile. We work around
+   * that by only doing GC if the lockfile is "old". */
+  if (lock_fd < 0 ||
+      fcntl (lock_fd, F_SETLK, &non_exclusive_lock) != 0)
```

Comment:
This should be `F_SETLKW`. If a concurrent process is currently doing GC, we should wait for it instead of just failing.